src/StreamJsonRpc/Reflection/MessageFormatterProgressTracker.cs

```diff
@@ -0,0 +1,168 @@
+        }
+
+        /// <summary>
+        /// Saves a <see cref="ProgressParamInformation"/> instance obtained from the given <see cref="object"/> into <see cref="ProgressMap"/>.
+        /// </summary>
+        /// <param name="value">The object which should implement <see cref="IProgress{T}"/> to create the <see cref="ProgressParamInformation"/> instance.</param>
+        /// <returns>The assigned <see cref="long"/> progres ID.</returns>
+        public long AddProgressObjectToMap(object value)
+        {
+            Requires.NotNull(value, nameof(value));
+
+            if (this.RequestIdBeingSerialized == null)
+            {
+                throw new NotSupportedException("IProgress<T> objects should not be part of any response or notification.");
+            }
+
+            lock (this.ProgressLock)
+            {
+                long progressId = this.NextProgressId++;
+                this.RequestProgressMap.Add(this.RequestIdBeingSerialized.Value, progressId);
+
+                this.ProgressMap.Add(progressId, new ProgressParamInformation(value));
+
+                return progressId;
+            }
+        }
+
+        /// <summary>
+        /// Class used to keep relevant information of an object that implements <see cref="IProgress{T}"/>.
+        /// </summary>
+        public class ProgressParamInformation
+        {
+            /// <summary>
+            /// Initializes a new instance of the <see cref="ProgressParamInformation"/> class.
+            /// </summary>
+            /// <param name="progressObject">The object implementing <see cref="IProgress{T}"/>.</param>
+            public ProgressParamInformation(object progressObject)
+            {
+                Requires.NotNull(progressObject, nameof(progressObject));
+
+                Type iProgressOfTType = MessageFormatterProgressTracker.FindIProgressOfT(progressObject.GetType());
+
+                Verify.Operation(iProgressOfTType != null, Resources.FindIProgressOfTError);
+
+                this.ValueType = iProgressOfTType.GenericTypeArguments[0];
+                this.ReportMethod = iProgressOfTType.GetRuntimeMethod(nameof(IProgress<int>.Report), new Type[] { this.ValueType });
+                this.ProgressObject = progressObject;
+            }
+
+            /// <summary>
+            /// Gets the actual <see cref="Type"/> reported by <see cref="IProgress{T}"/>.
+            /// </summary>
+            public Type ValueType { get; }
+
+            /// <summary>
+            /// Gets the <see cref="MethodInfo"/> of <see cref="IProgress{T}.Report(T)"/>.
+            /// </summary>
+            public MethodInfo ReportMethod { get; }
+
+            /// <summary>
+            /// Gets the instance of the object implementing <see cref="IProgress{T}"/>.
+            /// </summary>
+            public object ProgressObject { get; }
+        }
```

Comment:
Instead of exposing these two properties, can we make them private and offer the formatter just one method that will invoke the method on the caller's behalf?
That way, in the future if we optimize the code (e.g. reuse the one-element array that reflection requires) it will benefit all the formatters.